### ISSUE TITLE
Does WebSocketApp.run_forever() respond to pings?

### ISSUE THREAD

**Issue by Julian-O** [websocket-client/websocket-client#730]
The [documentation](https://websocket-client.readthedocs.io/en/latest/examples.html#post-connection-features) states:

> The `run_forever()` function will automatically send a “pong” when it receives a “ping”, per the specification.

Terrific. This is the functionality I would like for free.

But when I have tracing turned on, I can see the remote server sending me pings, and no evidence of a pong being returned. 

[To avoid confusion: I am talking about responding the pings initiated by the remote end. Not initiating pings to ensure the remote end is present.]

So, I hunted through the code, looking for where it responds, and I cannot see it. It calls-back `on_ping()`, if provided, but doesn't automatically pong. 

Am I missing something? I don't know whether the functionality is missing, or I am having a bad day reading code.

I would have expected that, if `on_ping()` is not provided by the user, a default callback function would be used that performed the send the pong, but if `on_ping()` is provided it becomes the responsibility of the end-client to respond, so that they may write custom code that verifies more of the stack is functioning.

(Testing with V1.2.1.)

**Comment by Julian-O** [comment]
> I can see the remote server sending me pings, and no evidence of a pong being returned.

Well, upon further investigation, the pings being sent by the third-party server are using `OPCODE_TEXT`, rather than `OPCODE_PING`, so they should **not** be automatically responded to. My question still stands though.

**Comment by engn33r** [comment]
Looks like you're right. I suspect [this line](https://github.com/websocket-client/websocket-client/blob/5f32b3c0cfb836c016ad2a5f6caeff2978a6a16f/websocket/_app.py#L338), currently line 338 of _app.py, is where the ping should get automatically handled. Unfortunately I don't think any on_ping() callback function exists by default, so it looks like either the documentation and/or the code needs an update. Feel free to submit a PR if you have a fix in mind, otherwise I'll get around to this later.

As you pointed out, a server using `OPCODE_TEXT` for a ping would not trigger the on_ping() callback function.

**Comment by Julian-O** [comment]
I don't need this functionality at the moment, because the server I am talking to is non-conformant anyway, but I am considering what would be required to make the change anyway to give back tot he community.

Here is my mulling out aloud.

# Code changes

* Define a public `send_pong()` method that transmits the appropriate OpCode, using send to do the work. Include comment that it is not necessary to call this, unless an `on_ping` parameter is provided.
* Define a private `_on_ping()` callback method that calls `send_pong()`
* Change `\_\_init\_\_()` to say `self.on_ping = on_pong or self._on_ping()`
* Update comments on `\_\_init\_\_()`.

# Documentation
* Explain in `examples.rst` that `run_forever()` will automatically respond to pings... unless you provide an `on_ping` callback. You may like to do that if you client's availability is dependent on other services, and only wants to reply with a pong message if those services are also available. In that case, the helper method `send_pong()` will be useful.
* Consider moving the description out of `examples.rst` - might not be the best place to explain API features.

# Verification (Does my code work as I expect?)
* This is where I start to get a bit worried this is going to be difficult. We need a new local server - similar to echo server - that sends pings and fails if no pongs are received. How will it indicate failure?

# Validation (Does what I thought was necessary actually work with servers in the wild?)
* This is where we need examples of a public server that emits pings and will disconnect if it doesn't get pong responses. I don't have the experience with WebSockets to be able to point to any of those. I suspect they are not common given the project hasn't encountered a problem until now.

Looking at this list, I don't think my skill set is going to be useful. I am confident I can write the software, but not confident I can get it tested it.

**Comment by lukecyca** [comment]
Is this a regression? I am running version 0.59.0 in production, and `run_forever()` appears to be automatically sending a PONG for every PING it receives from the server.

**Comment by Julian-O** [comment]
> Is this a regression?

I spot-checked a few older versions, and I could not see in any of them any code to respond to a ping with a pong. (I am too unfamiliar with the code to make an absolute statement here.)

**Comment by lukecyca** [comment]
I upgraded to the latest version (1.2.1) in a dev environment today and can confirm that it responds to pings. The server sends a PING, and the client running this library responds with a PONG.

I can't speak to how this library works as I'm not at all familiar. I just happen to be troubleshooting a different issue with ping/keepalive and found this issue and figured I'd give you my datapoint. Good luck!

**Comment by engn33r** [comment]
Like @lukecyca says, I just tested this and websocket-client responds as expected by the spec. If you want to run the same test that I did, do the following:
1. use the echo-server.py file (under websocket/tests/) and change the line with the serve function to:
'async with websockets.serve(echo, "localhost", LOCAL_WS_SERVER_PORT, ping_interval=2):'
2. Add the [4 lines of debug code](https://websockets.readthedocs.io/en/stable/cheatsheet.html#debugging) to echo-server.py. Change the logging level from `logging.INFO` to `logging.DEBUG` to see the outputs in your terminal
3. Use the [WebSocketApp ping/pong example](https://websocket-client.readthedocs.io/en/latest/examples.html#ping-pong-usage) with the proper IP and port (127.0.0.1:8765 by default) and you will see that websocket-client receives a ping every 2 seconds and the server receives a pong each time in the server log output.

@Julian-O let us know if you're not able to replicate this or if you had a different setup in mind. I'll write a similar check to the above test as a unit test for CI at some point in the future.

**Comment by Julian-O** [comment]
This initially confused me a lot. (The code is *haunted*! It is replying to pings, but there is no code to tell it to do so!) So I basically walked through the entire code-base until I found it.

In `websocket/_core.py`, line 415, `WebSocket.recv_data_frame()` sends the pong.

Is this the Right Place™ to put it? My first reaction is it is too low-level. It doesn't allow a client to override the ping behaviour. (e.g. if you are implementing a web-socket relay, you might want to pass the ping to the next server along, and pong only when it pongs.) But I am clearly very new to this technology, and I don't think my opinion should carry much weight.

So, the code is working as designed.

But the documentation is wrong, and it lead me up a garden path.

Under examples, it explicitly says about `WebSocket`:

> This type of connection does not automatically respond to a “ping” with a “pong”.

Under WebSocketApp, it says  `run_forever()`

> will automatically send a “pong” when it receives a “ping”, per the specification.

I hope to get a PR to you over the next couple of days to address the documentation error.

[Just for clarity: The server I use was sending non-conformant pings and was (quite rightly!) not getting the non-conformant responses it was expecting. This was what initially set me off down this path. I have not seen any behaviour that contradicts what you report.]

**Comment by engn33r** [comment]
Many thanks for diving into the code and figuring this out quicker than I did! The PR looks good, I expect to merge it in a day or two. I am open to ideas on moving this code from `recv_data_frame()` to other locations, but so far I have not seen any user requests for customizing the ping/pong feature at such a low level.

Lastly, a clarification: I realized the example code I pointed to in step 3 of my previous comment defined a `on_ping()` function, which as discussed before might enable the pong response to a ping that might not otherwise happen. After removing the custom `on_ping()` function from the client code, I confirmed that websocket-client still responds to pings from the server with a pong, without needing a custom `on_ping()` function.
